src/lightning/pytorch/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,488 @@
+
+        # temporary solution since state_dict will include
+        # model.state_dict(), model.get_shard_metadata(), and optimizer.state_dict()
+        converted_state = checkpoint.pop("state_dict")
+
+        import torch_xla.core.xla_model as xm
+
+        # ensure model parameters are updated
+        xm.mark_step()
+
+        parallel_devices = self.parallel_devices
+        assert parallel_devices is not None
+        if self._sequential_save:
+            # each host runs this in parallel, but the ranks in the host run it sequentially
+            for rank in range(len(parallel_devices)):
+                if rank == self.local_rank:
+                    self._save_checkpoint_shard(path, converted_state, storage_options)
+                self.barrier(f"wait-for-{rank}-save")
+        else:
+            self._save_checkpoint_shard(path, converted_state, storage_options)
+
+        if self._state_dict_type == "full":
+            ckpt_prefix = str(path / "checkpoint")
+            ckpt_suffix = "_rank-*-of-*.pth"
+            if len(parallel_devices) != self.world_size:  # multihost
+                raise OSError(
+                    "Multihost setups do not have a shared filesystem, so the checkpoint shards cannot be consolidated"
+                    " into a single checkpoint after saving them. Please switch to"
+                    " `XLAFSDPStrategy(state_dict_type='sharded')`. TIP: You can consolidate them manually by getting"
+                    " them together into a single directory and running `python -m"
+                    f" torch_xla.distributed.fsdp.consolidate_sharded_ckpts --ckpt_prefix {ckpt_prefix!r} --ckpt_suffix"
+                    f" {ckpt_suffix!r} --save_path 'path/to/consolidated.ckpt'`."
+                )
+
+            from torch_xla.distributed.fsdp import consolidate_sharded_model_checkpoints
+
+            self.barrier("before_ckpt_consolidation")
+            if self.is_global_zero:
+                save_path = path.parent / "consolidated.ckpt"
+                # save consolidated checkpoint separate to the shards
+                consolidate_sharded_model_checkpoints(ckpt_prefix, ckpt_suffix, str(save_path))
+                # remove the shards directory
+                self.checkpoint_io.remove_checkpoint(path)
+            self.barrier("after_ckpt_consolidation")
+
+    def _save_checkpoint_shard(
+        self,
+        path: Path,
+        converted_state: Dict[str, Union[Module, Optimizer, Any]],
+        storage_options: Optional[Any],
+    ) -> None:
+        self.checkpoint_io.save_checkpoint(
+            converted_state,
+            path / f"checkpoint_rank-{self.global_rank:08d}-of-{self.world_size:08d}.pth",
+            storage_options=storage_options,
+        )
+
+    def load_checkpoint(
+        self,
+        path: _PATH,
+        strict: bool = True,
+    ) -> Dict[str, Any]:
+        raise NotImplementedError("`load_checkpoint` is not implemented for `XLAFSDPStrategy.")
```

Comment:
I'm not sure that this should be merged without supporting checkpoint loading. Is this because it has to be loaded inside the Trainer?

Reply:
Yes. The regular FSDP code, when loading trainer, will use `self.model` and `self.lightning_module` as seen in https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L552-L553.

However, we don't have access to these.